haystack/components/evaluators/document_map.py

```diff
@@ -69,24 +69,21 @@ def run(
 
         for ground_truth, retrieved in zip(ground_truth_documents, retrieved_documents):
             score = 0.0
-            for ground_document in ground_truth:
-                if ground_document.content is None:
-                    continue
-
-                average_precision = 0.0
-                relevant_documents = 0
+            average_precision = 0.0
+            relevant_documents = 0
 
-                for rank, retrieved_document in enumerate(retrieved):
-                    if retrieved_document.content is None:
-                        continue
+            ground_truth_content = [doc.content for doc in ground_truth if doc.content is not None]
+            for rank, retrieved_document in enumerate(retrieved):
+                if retrieved_document.content is None:
+                    continue
 
-                    if ground_document.content in retrieved_document.content:
-                        relevant_documents += 1
-                        average_precision += relevant_documents / (rank + 1)
-                if relevant_documents > 0:
-                    score = average_precision / relevant_documents
+                if retrieved_document.content in ground_truth_content:
+                    relevant_documents += 1
+                    average_precision += relevant_documents / (rank + 1)
+            if relevant_documents > 0:
+                score = average_precision / relevant_documents
             individual_scores.append(score)
 
-        score = sum(individual_scores) / len(retrieved_documents)
+        score = sum(individual_scores) / len(ground_truth_documents)
```

Comment:
Not your fault but I had a hard time understanding the algorithm.
- I think we could benefit from a better choice of variable names.
- Adding a comment in the code with the formula/link from which we drew inspiration could also help

Resources I took inspiration from: [#1](https://www.pinecone.io/learn/offline-evaluation/#Mean-Average-Precision-(MAP)), [#2](https://amitness.com/posts/information-retrieval-evaluation#average-precisionap)

Something like this:

```python
            average_precision_numerator = 0.0
            relevant_documents = 0

            ground_truth_contents = [doc.content for doc in ground_truth if doc.content is not None]
            for rank, retrieved_document in enumerate(retrieved):
                if retrieved_document.content is None:
                    continue

                if retrieved_document.content in ground_truth_contents:
                    relevant_documents += 1
                    precision_at_k= relevant_documents / (rank + 1)
                    average_precision_numerator+=precision_at_k
            
            average_precision=0
            if relevant_documents > 0:
                average_precision = average_precision_numerator / relevant_documents
            individual_scores.append(average_precision)

        score = sum(individual_scores) / len(ground_truth_documents)
```
(untested)

WDYT?

Reply:
Agreed, I also struggled to understand the function initially. Optimized variables can help.
 Not sure if we need to add the link to the formula as its pretty common.

Reply:
Also I think `precision_at_k` is unnecessary and will be extra storage?

Reply:
I agree that it is not necessary, but in my opinion it makes the code more readable.

Feel free to make any changes you think appropriate.